### PR TITLE
[SPARK-41811][PYTHON][CONNECT] Implement `SQLStringFormatter` with `WithRelations`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement `SQLStringFormatter` for Python Client

### Why are the changes needed?
for parity

### Does this PR introduce _any_ user-facing change?
yes, new feature

```
In [1]: mydf = spark.range(10)

In [2]: spark.sql("SELECT {col} FROM {mydf} WHERE id IN {x}", col=mydf.id, mydf=mydf, x=tuple(range(4))).show()
+---+                                                                           
| id|
+---+
|  0|
|  1|
|  2|
|  3|
+---+
```

### How was this patch tested?
enabled doc tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -131,6 +132,23 @@ message SQL {
...
 }
 
+// Relation of type [[WithRelations]].
+//
+// This relation contains a root plan, and one or more references that are used by the root plan.
+// There are two ways of referencing a relation, by name (through a subquery alias), or by plan_id
+// (using RelationCommon.plan_id).
+//
+// This relation can be used to implement CTEs, describe DAGs, or to reduce tree depth.
+message WithRelations {
```
Comment (xinrong-meng) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:142]: May I ask which APIs are associated with the relation?

**Review by HyukjinKwon — APPROVED** [review]
Seems fine to me but would need @hvanhovell's look

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (zhengruifeng) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:142]: @hvanhovell would you mind help review this one? thanks

**Comment by HyukjinKwon** [comment]
Let's just go ahead as all comments are addresed.

File: python/pyspark/sql/session.py [hunk 2]
```diff
@@ -1662,7 +1662,7 @@ def sql(
 
         And substitute named parameters with the `:` prefix by SQL literals.
 
-        >>> from pyspark.sql.functions import create_map
+        >>> from pyspark.sql.functions import create_map, lit
```
Comment (xinrong-meng) [inline python/pyspark/sql/session.py:1665]: Out of curiosity, why is the missing import not detected? Are we skipping the doctest?

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
LGTM after resolving conflicts, thanks!

File: python/pyspark/sql/session.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/session.py:1665]: oh, `lit` was import at the beginning of this file.
but for the parity doctest, it is missing, so add it here

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by zhengruifeng** [comment]
thanks @HyukjinKwon and @xinrong-meng

File: python/pyspark/sql/session.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/sql/session.py:1698]: Thanks!

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 3]
```diff
@@ -61,7 +61,7 @@ message Command {
 // almost oblivious to the server-side behavior.
 message SqlCommand {
   // (Required) SQL Query.
-  string sql = 1;
+  string sql = 1 [deprecated=true];
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:64]: @zhengruifeng where am I supposed to set the sql text now?

Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:64]:
```
  // (Optional) The relation that this SQL command will be built on.
  Relation input = 6;
```

this Relation can be a `message SQL` or `message WithRelations`

**Pull request closed without merging**